services/wallet/api.go

```diff
@@ -874,6 +887,48 @@ func (api *API) SignTypedDataV4(typedJson string, address string, password strin
 	return types.HexBytes(sig), err
 }
 
+// SafeSignTypedDataForDApps is used to execute requests for "eth_signTypedData"
+// if legacy is true else "eth_signTypedData_v4"
+// the formatted typed data won't be prefixed in case of legacy calls, as the
+// old dApps implementation expects
+// the chain is validate for both cases
+func (api *API) SafeSignTypedDataForDApps(typedJson string, address string, password string, chainID uint64, legacy bool) (types.HexBytes, error) {
+	log.Debug("wallet.api.SafeSignTypedDataForDApps", "len(typedJson)", len(typedJson), "address", address, "len(password)", len(password), "chainID", chainID, "legacy", legacy)
+
+	account, err := api.getVerifiedWalletAccount(address, password)
+	if err != nil {
+		return types.HexBytes{}, err
+	}
+
+	var typed typeddata.TypedData
```

Comment:
This, lines `903` - `907` could go under "then"block of `if legacy {`.
@stefandunca

Reply:
The un-marshaing and following call to `ValidateChainID` is used in both branches of legacy call.